src/postprocess/chi0.jl

```diff
@@ -131,53 +166,77 @@ returns `3` extra bands, which are not converged by the eigensolver
     # not be necessary, but it simplifies the interaction with the
     # Sternheimer linear solver (it makes the rhs be order 1 even if
     # δV is small)
-    normδV = norm(δV.real)
-    normδV < eps(T) && return RealFourierArray(basis)
+    if n_spin == 1
+        normδV = norm(δV.real)
+        normδV < eps(T) && return (RealFourierArray(basis), )
+    else
+        normδV = sqrt(sum(abs2, δV.real) + sum(abs2, δVβ.real))
+        normδV < eps(T) && return (RealFourierArray(basis), RealFourierArray(basis))
+    end
 
     # Make δV respect the full model symmetry group, since it's
     # invalid to consider perturbations that don't (technically it
     # could be made to only respect basis.symmetries, but symmetrizing wrt
     # the model symmetry group means that χ0 is unaffected by the
     # use_symmetry kwarg of basis, which is nice)
-    δV = symmetrize(δV).real / normδV
+    δV = symmetrize(δV).real  / normδV
+    (n_spin > 1) && (δVβ = symmetrize(δVβ).real / normδV)
 
     if droptol > 0 && sternheimer_contribution == true
         error("Droptol cannot be positive if sternheimer contribution is to be computed.")
     end
 
+    # Accumulate spin component by spin component:
     # δρ = ∑_nk (f'n δεn |ψn|^2 + 2Re fn ψn* δψn - f'n δεF |ψn|^2
-    δρ_fourier = zeros(complex(T), size(δV))
-    for ik = 1:length(basis.kpoints)
+    δρ_fourier = [zeros(complex(T), size(δV)) for _ in 1:n_spin]
+    for (ik, kpt) in enumerate(basis.kpoints)
         δρk = zero(δV)
         for n = 1:size(ψ[ik], 2)
+            δV_spin_component = kpt.spin == 1 ? δV : δVβ
             add_response_from_band!(δρk, n, ham.blocks[ik], eigenvalues[ik], ψ[ik],
-                                    εF, δV, temperature, droptol, sternheimer_contribution,
-                                    kwargs_sternheimer)
+                                    εF, δV_spin_component, temperature, droptol,
+                                    sternheimer_contribution, kwargs_sternheimer)
         end
-        accumulate_over_symmetries!(δρ_fourier, r_to_G(basis, complex(δρk)),
+        accumulate_over_symmetries!(δρ_fourier[kpt.spin], r_to_G(basis, complex(δρk)),
                                     basis, basis.ksymops[ik])
     end
-    δρ = real(G_to_r(basis, δρ_fourier))
-    count = sum(length(basis.ksymops[ik]) for ik in 1:length(basis.kpoints))
-    δρ ./= count
+    count = sum(length(basis.ksymops[ik]) for ik in 1:length(basis.kpoints)) ÷ n_spin
+    δρ = [real(G_to_r(basis, δρσ)) ./ count for δρσ in δρ_fourier]
 
     # Add variation wrt εF
     if temperature > 0
         dVol = basis.model.unit_cell_volume / prod(basis.fft_size)
-        ldos = LDOS(εF, basis, eigenvalues, ψ, temperature=temperature)
+        ldos = [LDOS(εF, basis, eigenvalues, ψ, temperature=temperature, spins=[σ])
+                for σ in 1:n_spin]
         dos  = DOS(εF, basis, eigenvalues, temperature=temperature)
-        δρ .+= ldos .* dot(ldos, δV) .* dVol ./ dos
+        if n_spin == 1
```

Comment:
can be factorized by first computing the delta eps_f (spin-dependent part)

Reply:
Sorry, I don't get what you mean.

Reply:
delta eps_f = ...; nspin == 2 && delta eps_f += ...; delta rho .+= ldos .* delta eps_f .* ...

Reply:
Ah I see ... indeed makes it a bit nicer to read.